core/lib/zksync_core/src/state_keeper/cached_storage.rs

```diff
@@ -0,0 +1,270 @@
+use std::fmt::Debug;
+
+use anyhow::Context;
+use tokio::{runtime::Handle, sync::watch, task::JoinHandle};
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_state::{PostgresStorage, ReadStorage, RocksdbStorage};
+use zksync_types::L1BatchNumber;
+
+type BoxReadStorage<'a> = Box<dyn ReadStorage + Send + 'a>;
+
+/// Encapsulates a storage that can produce a short-lived [`ReadStorage`] implementation backed by
+/// either Postgres or RocksDB (if it's caught up). Maintains internal state of how behind RocksDB
+/// is compared to Postgres and actively tries to catch it up in the background.
+///
+/// This struct's main design purpose is to be able to produce [`ReadStorage`] implementation with
+/// as little blocking operations as possible to ensure liveliness.
+#[derive(Debug)]
+pub struct CachedStorage {
```

Comment:
I certainly agree with @RomanBrodetski in that this looks overly complex. Conceptually, we want to represent two cases: either the storage has caught up (=> use RocksDB) or not (=> use Postgres). We can assume that the storage doesn't fall behind unless we let it to. So this should roughly correspond to

```rust
#[derive(Clone)]
enum CachedStorageFactory {
    Postgres(ConnectionPool),
    Rocksdb(RocksDB<StateKeeperColumnFamily>),
}

impl CachedStorageFactory {
    pub async fn access_storage(&self) -> Box<dyn ReadStorage + Send + '_> {
        // trivial. We may even not box the output type, but rather have an enum for it
        // (theoretically, can lead to better performance due to more
        // compiler optimizations being possible, so I'd go for it).
    }
}

pub struct CachedStorage {
    inner: Arc<Mutex<CachedStorageFactory>>,
    // ^ the background update task will replace `inner` once caught up.
}

impl CachedStorage {
    fn factory(&self) -> CachedStorageFactory {
         // it's important that we don't hold the lock for long;
         // that's why `CachedStorageFactory` implements `Clone`
         self.inner.lock().expect("poisoned").clone()
    }
}
```

- With this approach, an update task can be treated in the same way as other server / EN tasks; you wouldn't need to retain its handle locally.
- If you'd want to retain `ConnectionPool` in `CachedStorageFactory` even after it caught up (e.g., for it to be able to update to the latest L1 batch), this is possible as well by slightly changing its shape.

Reply:
Indeed, I tried to account for the possibility that RocksDB somehow falls behind Postgres in the middle of the execution but if we don't care about that it simplifies a lot of things.

I have tried out your approach and it works perfectly. Just wondering what do you think I should do about errors? The existing codebase seems to be pretty split between:
a. Just unwrapping
b. Using `anyhow` for everything
c. Proper structured errors.

I feel like structured errors could make some sense here even if most underlying code is just `anyhow`. Also, representing interruption via an error would look cleaner than `None` as it is now (inherited approach from `RocksdbStorage`).